test-app/src/main/java/com/mapbox/navigation/examples/core/MapboxManeuverActivity.kt

```diff
@@ -0,0 +1,384 @@
+import com.mapbox.navigation.base.options.NavigationOptions
+import com.mapbox.navigation.base.trip.model.RouteProgress
+import com.mapbox.navigation.core.MapboxNavigation
+import com.mapbox.navigation.core.directions.session.RoutesObserver
+import com.mapbox.navigation.core.directions.session.RoutesRequestCallback
+import com.mapbox.navigation.core.internal.formatter.MapboxDistanceFormatter
+import com.mapbox.navigation.core.replay.MapboxReplayer
+import com.mapbox.navigation.core.replay.ReplayLocationEngine
+import com.mapbox.navigation.core.replay.route.ReplayProgressObserver
+import com.mapbox.navigation.core.replay.route.ReplayRouteMapper
+import com.mapbox.navigation.core.trip.session.BannerInstructionsObserver
+import com.mapbox.navigation.core.trip.session.LocationObserver
+import com.mapbox.navigation.core.trip.session.RouteProgressObserver
+import com.mapbox.navigation.examples.util.Slackline
+import com.mapbox.navigation.examples.util.Utils
+import com.mapbox.navigation.ui.base.api.maneuver.ManeuverApi
+import com.mapbox.navigation.ui.base.api.maneuver.ManeuverCallback
+import com.mapbox.navigation.ui.base.api.maneuver.StepDistanceRemainingCallback
+import com.mapbox.navigation.ui.base.api.maneuver.UpcomingManeuversCallback
+import com.mapbox.navigation.ui.base.model.maneuver.ManeuverState
+import com.mapbox.navigation.ui.maneuver.api.MapboxManeuverApi
+import com.mapbox.navigation.ui.utils.internal.ifNonNull
+import kotlinx.android.synthetic.main.layout_activity_maneuver.*
+import java.lang.ref.WeakReference
+import java.util.Objects
+
+class MapboxManeuverActivity : AppCompatActivity(), OnMapLongClickListener {
+
+    private lateinit var mapboxMap: MapboxMap
+    private lateinit var mapCamera: CameraAnimationsPlugin
+    private lateinit var mapboxNavigation: MapboxNavigation
+    private lateinit var maneuverApi: ManeuverApi
+    private var locationComponent: LocationPluginImpl? = null
+
+    private val mapboxReplayer = MapboxReplayer()
+    private val replayRouteMapper = ReplayRouteMapper()
+    private val slackLine = Slackline(this)
+    private val locationEngineCallback = MyLocationEngineCallback(this)
+    private val replayProgressObserver = ReplayProgressObserver(mapboxReplayer)
+
+    private val currentManeuverCallback = object : ManeuverCallback {
+        override fun onManeuver(currentManeuver: ManeuverState.CurrentManeuver) {
+            val maneuver = currentManeuver.maneuver
+            val primary = maneuver.primary
+            val secondary = maneuver.secondary
+            val sub = maneuver.sub
+            val lane = maneuver.laneGuidance
+            if (secondary?.componentList != null) {
+                maneuverView.render(ManeuverState.ManeuverSecondary.Show)
+                maneuverView.render(ManeuverState.ManeuverSecondary.Instruction(secondary))
+            } else {
+                maneuverView.render(ManeuverState.ManeuverSecondary.Hide)
+            }
```

Comment:
Why is this decision not made by the API but instead consumer has to process the data?

Reply:
API is only concerned with returning back the data and knows nothing about the state of UI or when to show/hide. So, in this case either there are `secondary` instructions or not. It's up to the user to decide what they want to show on the UI based on the availability of the `secondary` instructions. If you want to use `Maneuvers` as a standalone widget, you have to handle it. If you don't want then use the `DropIn UI`.

Reply:
To me this makes sense because if the API class is too opinionated about how the data is displayed it makes it harder for a developer to customize their own view component. As Abhishek points out the API should return data resulting from the call that was made but know have too much knowledge about how the data is being displayed.

Reply:
That makes sense but again is different than what we did with other components. As an example, `MapboxSnapshotView` does not expose separate states to hide/show, and the route line has the `UpdateLayerVisibilityState` which is actually produced by the API.

This means that our current implementation has 3 variants of the approach currently implemented in the SDK:
- pass command directly to view
- do not pass anything to view and let it do its thing
- use API to generate visibility state and pass it to view

Let's make sure that we discuss this today to align.